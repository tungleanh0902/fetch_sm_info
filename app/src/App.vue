<template>
  <div id="app" :class="themeClass">
    <NetworkSelector />
    <AppHeader />
    <div id="app-content">
      <div id="bar-container">
        <CookieBar v-if="!isMobileApp" />
        <MaintenanceBar />
        <DisconnectedBar />
      </div>
      <UserMenu v-if="!isMobileApp" />
      <router-view name="session" />
      <router-view />
    </div>
    <MobileMenu />
  </div>
</template>

<script>
import { mapState, mapGetters } from "vuex"
import AppHeader from "common/AppHeader"
import CookieBar from "common/CookieBar"
import DisconnectedBar from "common/DisconnectedBar"
import UserMenu from "account/UserMenu"
import MaintenanceBar from "common/MaintenanceBar"
import MobileMenu from "common/MobileMenu"
import NetworkSelector from "common/NetworkSelector"
import store from "./vuex/store"

export default {
  name: `app`,
  components: {
    AppHeader,
    CookieBar,
    DisconnectedBar,
    UserMenu,
    MaintenanceBar,
    MobileMenu,
    NetworkSelector,
  },
  computed: {
    ...mapState([`session`]),
    ...mapGetters([`network`]),
    isMobileApp() {
      return this.session.mobile
    },
    themeClass() {
      return !this.$route.meta.networkSpecificRoute
        ? `lunie-light`
        : this.network
    },
  },
  store,
}
</script>

<style>
@import "./styles/app.css";
</style>
