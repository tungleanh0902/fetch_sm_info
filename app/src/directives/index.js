export const focusElement = {
  inserted: (el) => {
    el.focus()
  },
}

// Directive to Focus an element, but only if a child does not have focus
export const focusParentLast = {
  inserted: (el) => {
    if (!el.contains(document.activeElement)) {
      el.focus()
    }
  },
}
