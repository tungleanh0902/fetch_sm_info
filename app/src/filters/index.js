import { percent } from "../scripts/num"
import moment from "moment"
import bech32 from "bech32"

export const date = (date) => moment(date).format("MMMM Do YYYY, HH:mm")

export const fromNow = (date) => moment(date).fromNow()

export const noBlanks = function (value) {
  return value === undefined ||
    value === null ||
    value === `` ||
    value === `[do-not-modify]`
    ? `--`
    : value
}

export const percentOrPending = function (value, totalValue, pending) {
  return pending ? `--` : percent(totalValue === 0 ? 0 : value / totalValue)
}

const getAddressType = (address) => {
  if (address.startsWith("0x")) return "ethereum"
  try {
    bech32.decode(address)
    return "cosmos"
  } catch (error) {
    // ignore error
  }
  return "any"
}

export const formatAddress = (address, length = 4) => {
  if (!address) {
    return `Address Not Found`
  }
  switch (getAddressType(address)) {
    case "cosmos":
      return address.split(`1`)[0] + `…` + address.slice(-1 * length)
    case "ethereum":
      return address.slice(0, 2 + length) + `…` + address.slice(-1 * length)
    case "any":
      return address.slice(0, length) + `…` + address.slice(-1 * length)
  }
}

export const resolveValidatorName = (address, validators) => {
  // Substrate validators can have their operatorAddress as their name, which is too long and thus breaks the
  // extension size. If name is too long, then we just display the formatted address
  if (validators[address] && validators[address].name.length < 20) {
    return validators[address].name
  }
  return formatAddress(address)
}

export const validatorEntry = (validator) =>
  `${validator.name} - ${formatAddress(validator.operatorAddress, 20)}`
