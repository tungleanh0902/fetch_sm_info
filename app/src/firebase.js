import config from "../config"
import * as Sentry from "@sentry/browser"

// load firebase async
let firebaseInstance
const getFirebase = async () => {
  if (firebaseInstance) {
    return firebaseInstance
  }
  try {
    const [firebase] = await Promise.all([
      import("firebase/app"),
      import("firebase/auth"),
    ])
    firebase.initializeApp(config.firebaseConfig)
    firebaseInstance = firebase
    return firebase
  } catch (error) {
    console.error(`Firebase could not be initialized`)
    console.error(error)
    Sentry.captureException(error)
  }
}

export default getFirebase
