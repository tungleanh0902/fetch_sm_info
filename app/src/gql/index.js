/* istanbul ignore file */

import gql from "graphql-tag"
import store from "../vuex/store"

function getCurrentNetwork() {
  return store().state.connection.network
}

export const schemaMap = {
  cosmoshub: "",
  [`gaia-testnet`]: "gaia_testnet_",
  testnet: "gaia_testnet_",
}

export const ValidatorFragment = `
  id
  name
  operatorAddress
  consensusPubkey
  jailed
  picture
  details
  website
  identity
  votingPower
  startHeight
  uptimePercentage
  tokens
  commissionUpdateTime
  commission
  maxCommission
  maxChangeCommission
  status
  statusDetailed
  expectedReturns
  selfStake
`

export const AllValidators = () => {
  const currentNetwork = getCurrentNetwork()
  return gql`
    query AllValidators {
      validators(networkId: "${currentNetwork}") {
        ${ValidatorFragment}
      }
    }`
}

export const ValidatorProfile = gql`
  query validator($networkId: String!, $operatorAddress: String!) {
    validator(networkId: $networkId, operatorAddress: $operatorAddress) {
      ${ValidatorFragment}
    }
  }
`

export const DelegatorValidators = (schema) => gql`
  query ValidatorInfo($delegatorAddress: String!) {
    validators(networkId: "${schema}", delegatorAddress: $delegatorAddress) {
      ${ValidatorFragment}
    }
  }
`

export const DelegationsForDelegator = (schema) => gql`
  query delegations($delegatorAddress: String!) {
    delegations(networkId: "${schema}", delegatorAddress: $delegatorAddress) {
      id
      validator {
        ${ValidatorFragment}
      }
      amount
    }
  }
`

export const Networks = gql`
  query Networks {
    networks {
      id
      chain_id
      testnet
      title
      icon
      slug
      powered {
        name
        providerAddress
        picture
      }
    }
  }
`

// load all the data immediatly to avoid async loading later
export const NetworksAll = gql`
  query Networks($experimental: Boolean) {
    networks(experimental: $experimental) {
      id
      chain_id
      testnet
      title
      icon
      slug
      default
      lockUpPeriod
      powered {
        name
        providerAddress
        picture
      }
      feature_session
      feature_portfolio
      feature_validators
      feature_proposals
      feature_activity
      feature_explorer
      action_send
      action_claim_rewards
      action_delegate
      action_redelegate
      action_undelegate
      action_deposit
      action_vote
      action_proposal
      stakingDenom
      network_type
      address_creator
      address_prefix
      bech32_prefix
      ledger_app
      testnet
      enabled
      coinLookup {
        chainDenom
        viewDenom
        chainToViewConversionFactor
        icon
      }
      rpc_url
      HDPaths
      curves
      defaultHDPath
      defaultCurve
    }
  }
`

export const NetworksResult = (data) => data.networks

export const ProposalFragment = `
  id
  proposalId
  type
  title
  description
  creationTime
  status
  statusBeginTime
  statusEndTime
  tally {
    yes
    no
    veto
    abstain
    total
    totalVotedPercentage
  }
  deposit
  proposer {
    name
    address
    picture
  }
  validator {
    id
    name
    details
    identity
    picture
  }
  beneficiary {
    name
    address
    picture
  }
  summary
  detailedVotes {
    deposits {
      amount {
        amount
        denom
      }
      depositer {
        name
        address
        picture
      }
    }
    depositsSum
    percentageDepositsNeeded
    votes {
      id
      voter {
        name
        address
        picture
      }
      option
      amount {
        amount
        denom
      }
    }
    votesSum
    votingThresholdYes
    votingThresholdNo
    votingPercentageYes
    votingPercentageNo
    links {
      title
      link
      type
    }
    timeline {
      title 
      time
    }
  }
`

export const ProposalItem = (schema) => gql`
  query proposal($id: String!) {
    proposal(networkId: "${schema}", id: $id) {
      ${ProposalFragment}
    }
  }
`

export const GovernanceParameters = (schema) => gql`
query governanceParameters {
  governanceParameters(networkId: "${schema}") {
    depositDenom
    votingThreshold
    vetoThreshold
    depositThreshold
  }
}
`

export const Vote = (schema) => gql`
query vote($proposalId: String!, $address: String!) {
  vote(networkId: "${schema}", proposalId: $proposalId, address: $address) {
    option
  }
}
`

export const Block = (networkId) => gql`
query Block {
  block(networkId: "${networkId}") {
    id
    height
    chainId
  }
}
`

export const MetaData = (schema) => gql`
query metaData {
  metaData(networkId: "${schema}") {
    stakingDenom
  }
}
`

export const UserTransactionAdded = gql`
  subscription($networkId: String!, $address: String!) {
    userTransactionAddedV2(networkId: $networkId, address: $address) {
      hash
      height
      success
      log
    }
  }
`

export const NotificationAdded = gql`
  subscription($addressObjects: [NotificationInput]!) {
    notificationAdded(addressObjects: $addressObjects) {
      networkId
      timestamp
      title
      link
      icon
    }
  }
`

export const AddressRole = gql`
  query($networkId: String!, $address: String!) {
    accountRole(networkId: $networkId, address: $address)
  }
`
