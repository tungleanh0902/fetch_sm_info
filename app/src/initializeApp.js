/* istanbul ignore file: really just integrations */

import { Plugins } from "@capacitor/core"
import * as Sentry from "@sentry/browser"
const { App: CapacitorApp } = Plugins
import { listenToExtensionMessages } from "scripts/extension-utils"
import { checkForNewLunieVersions } from "scripts/check-for-new-lunie-versions"
import {
  enableGoogleAnalytics,
  setGoogleAnalyticsPage,
} from "scripts/google-analytics"
import config from "src/../config"
import Router, { routeGuard } from "./router"
import Store from "./vuex/store"
import { createApolloProvider, apolloErrorHandler } from "src/gql/apollo.js"
import { registerForPushNotifications } from "./scripts/pushNotifications"
import { handleDeeplink, getLaunchUrl } from "./vuex/modules/account"

if (navigator && navigator.serviceWorker) {
  // remove any existing service worker
  navigator.serviceWorker.getRegistrations().then(function (registrations) {
    for (let registration of registrations) {
      registration.unregister()
    }
  })
}

function setOptions(urlParams, store) {
  if (urlParams.experimental) {
    store.commit(`setExperimentalMode`)
  }
  if (urlParams.insecure === `true`) {
    store.commit(`setInsecureMode`)
  }
  if (urlParams.network) {
    store.dispatch(`setNetwork`, { id: urlParams.network })
  }
}

export const bootError = (error) => {
  document.querySelector("#app").innerHTML = `
  <div class="boot-error">
    <h2>Uh oh. Lunie failed to start. Our team has been notified of the issue. Please try refreshing this page in a few minutes. If the error persists please contact us at http://help.lunie.io/</h2>

    <p class="error">${error}</p>
  </div>
    `
}

export default async function init(urlParams, env = process.env) {
  let initialized = false
  // add error handlers in production
  if (env.NODE_ENV === `production`) {
    // check every minute if new Lunie versions have been deployed
    checkForNewLunieVersions()
    enableGoogleAnalytics(config.google_analytics_uid)
  }

  console.log(`Expecting backend at: ${config.graphqlHost}`)

  const apolloProvider = await createApolloProvider()
  const apolloClient = apolloProvider.clients.defaultClient

  const store = Store({ apollo: apolloClient })

  // we need to use this custom error handler as we want to use the store in there
  // we can't pass the store as it would create a circular dependency
  apolloErrorHandler.onError((error) => {
    if (error.extensions && error.extensions.code === "UNAUTHENTICATED") {
      store.dispatch("signOutUser")
      return
    }
    Sentry.captureException(error)
    console.error(error)

    if (!initialized) {
      bootError(new Error("Failed to connect to API: " + error.message))
    }
  })

  // we need to set url params before querying for networks because of experimental flag
  setOptions(urlParams, store)

  // check if user is signed in
  store.dispatch(`listenToAuthChanges`) // handles Google OAuth changes we are not influencing
  await store.dispatch(`checkSession`)

  // we load the networks first as we need them in the router
  await store.dispatch(`preloadNetworkCapabilities`)

  const router = Router(store)
  setGoogleAnalyticsPage(router.currentRoute.path)
  router.beforeEach(routeGuard(store))
  router.afterEach((to) => {
    /* istanbul ignore next */
    setGoogleAnalyticsPage(to.path)
  })

  registerForPushNotifications(store, router)

  if (config.mobileApp) {
    CapacitorApp.addListener("appUrlOpen", function (data) {
      handleDeeplink(data.url, router)
    })
    // handling deeplinks when app is opening
    getLaunchUrl(router)
  } else {
    handleDeeplink(window.location, router)
  }

  store.dispatch(`loadLocalPreferences`)
  await store.dispatch(`checkForPersistedNetwork`) // wait until signin
  store.dispatch(`checkForPersistedAddresses`)

  listenToExtensionMessages(store)

  initialized = true

  return { store, router, apolloProvider }
}
