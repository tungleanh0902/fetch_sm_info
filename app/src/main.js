/* istanbul ignore file */

import Vue from "vue"
import VTooltip from "v-tooltip"
import Vuelidate from "vuelidate"
import InfiniteScroll from "vue-infinite-scroll"
import VueClipboard from "vue-clipboard2"
import { DynamicReactiveRefs } from "vue-reactive-refs"
import { focusElement, focusParentLast } from "src/directives"
import App from "./App.vue"
import init, { bootError } from "./initializeApp"
import { getURLParams } from "scripts/url"
import { Plugins } from "@capacitor/core"
import config from "src/../config"
import * as Sentry from "@sentry/browser"
import * as Integrations from "@sentry/integrations"
import "material-design-icons-iconfont/dist/material-design-icons.css"
import AsyncComputed from "vue-async-computed"
import VueScrollTo from "vue-scrollto"

if (config.sentryDSN) {
  Sentry.init({
    dsn: config.sentryDSN,
    integrations: [new Integrations.Vue({ Vue, attachProps: true })],
    release: process.env.LUNIE_VERSION,
  })
}

Vue.config.productionTip = false

Vue.use(VTooltip)

Vue.use(Vuelidate)
Vue.use(VueClipboard)
Vue.use(InfiniteScroll)
Vue.use(AsyncComputed)
Vue.use(DynamicReactiveRefs)
Vue.use(VueScrollTo)

Vue.directive(`focus`, focusElement)
Vue.directive(`focus-last`, focusParentLast)

const urlParams = getURLParams(window)
init(urlParams)
  .then(({ store, router, apolloProvider }) => {
    const { SplashScreen, StatusBar } = Plugins

    new Vue({
      router,
      ...App,
      store,
      apolloProvider,
      mounted() {
        if (config.mobileApp) {
          SplashScreen.hide()
          StatusBar.show()
        }
      },
    }).$mount("#app")
  })
  .catch((error) => {
    Sentry.captureException(error)
    bootError(error)
  })
