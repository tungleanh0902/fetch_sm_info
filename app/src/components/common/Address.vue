<template>
  <div
    v-tooltip="tooltipText"
    class="copyable-address"
    :class="{
      'with-type': !!addressType,
    }"
  >
    <div
      v-clipboard:copy="address"
      v-clipboard:success="() => onCopy()"
      class="address"
    >
      <div class="address-section">
        <span v-if="addressType" class="type"> {{ addressType }} Address </span>
        <span>{{ address | formatAddress }}</span>
      </div>
      <div :class="{ active: copySuccess }" class="icon-container">
        <i class="material-icons notranslate success">check</i>
      </div>
      <i class="material-icons notranslate copy">content_copy</i>
    </div>
  </div>
</template>

<script>
import { formatAddress } from "src/filters"

export default {
  name: `lunie-address`,
  filters: {
    formatAddress,
  },
  props: {
    address: {
      type: String,
      required: true,
    },
    tooltipText: {
      type: String,
      default: ``,
    },
    addressType: {
      type: String,
      default: undefined,
    },
  },
  data: () => ({
    copySuccess: false,
  }),
  methods: {
    onCopy() {
      this.copySuccess = true
      setTimeout(() => {
        this.copySuccess = false
      }, 2500)
    },
  },
}
</script>
<style scoped>
.copyable-address {
  font-size: 14px;
  display: inline-flex;
  height: 2rem;
  padding: 0 1rem;
  border-radius: 1rem;
  margin: 0;
  background: var(--app-fg);
}

.copyable-address.with-type {
  height: 2.5rem;
}

.copyable-address .type {
  font-size: 10px;
  line-height: 10px;
  color: var(--dim);
  margin-top: 1px;
}

.copyable-address .address {
  white-space: nowrap;
  display: inline-flex;
  align-items: center;
  cursor: pointer;
  color: var(--link);
}

.copyable-address .address-section {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
}

.menu-address .address {
  color: var(--menu-link);
}

.copyable-address .address:hover {
  color: var(--link-hover);
}

.menu-address .address:hover {
  color: var(--menu-link-hover);
}

.copyable-address .icon-container {
  display: flex;
  align-items: center;
  justify-content: center;
}

.copyable-address .icon-container .success {
  opacity: 0;
  transition: opacity 250ms ease;
}

.copyable-address .icon-container.active .success {
  opacity: 1;
}

.copyable-address i {
  font-size: 14px;
  padding-left: 0.25rem;
}

.copyable-address .icon-container i.success {
  color: var(--success);
}

.material-icons {
  font-size: 14px;
}
</style>
