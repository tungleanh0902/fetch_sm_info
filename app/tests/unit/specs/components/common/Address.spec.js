import { shallowMount, createLocalVue } from "@vue/test-utils"
import VueClipboard from "vue-clipboard2"
import Address from "src/components/common/Address"
import { formatAddress } from "src/filters"

const localVue = createLocalVue()
localVue.directive(`clipboard`, VueClipboard)
localVue.directive(`tooltip`, () => {})
localVue.filter(`formatAddress`, formatAddress)

describe(`Address Component`, () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(Address, {
      localVue,
      propsData: { address: `cosmos1ek9cd8ewgxg9w5xllq9um0uf4aaxaruvcw4v9e` },
      data: () => ({
        copySuccess: false,
      }),
    })
  })

  it(`should show a short address`, () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`should return 'address not found'`, () => {
    wrapper.setProps({ address: null })
    expect(wrapper.find(".address").text()).toContain(`Address Not Found`)
  })

  it(`should return a short address with everything before the 1`, () => {
    wrapper.setProps({
      address: `cosmosvaloper1qecshyc40kshszkwrtscgmsdd8tz3n4hrj9yf2`,
    })
    expect(wrapper.find(".address").text()).toContain(`cosmosvaloperâ€¦9yf2`)
  })

  describe(`onCopy`, () => {
    it(`should set and reset copySuccess`, () => {
      jest.useFakeTimers()
      wrapper.vm.onCopy() // old test style to make timer work
      expect(wrapper.vm.copySuccess).toBe(true)

      jest.runAllTimers()
      expect(wrapper.vm.copySuccess).toBe(false)
    })
  })

  // TODO: not sure how to test the v-clipboard directive events
  xit(`clicking copy copies the address`, () => {
    jest.useFakeTimers()
    expect(wrapper.find(`.copied`).classes().includes(`active`)).toBe(false)

    wrapper.find(`.address`).trigger(`click`)
    expect(wrapper.find(`.copied`).classes().includes(`active`)).toBe(true)

    jest.runAllTimers()
    expect(wrapper.find(`.copied`).classes().includes(`active`)).toBe(false)
  })
})
