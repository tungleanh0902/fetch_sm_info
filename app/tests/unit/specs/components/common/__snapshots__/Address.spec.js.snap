// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Address Component should show a short address 1`] = `
<div
  class="copyable-address"
>
  <div
    class="address"
  >
    <div
      class="address-section"
    >
      <!---->
       
      <span>
        cosmosâ€¦4v9e
      </span>
    </div>
     
    <div
      class="icon-container"
    >
      <i
        class="material-icons notranslate success"
      >
        check
      </i>
    </div>
     
    <i
      class="material-icons notranslate copy"
    >
      content_copy
    </i>
  </div>
</div>
`;
